pkg/urbit/jets/e/en_json.c

```diff
@@ -0,0 +1,438 @@
+    }
+    if ( _(u3du(a)) ) {
+      return _encode_fail(buf_u);
+    }
+
+    c3_w len_w = u3r_met(3, a);
+
+    buf_u = _append_char(buf_u, '"');
+    for (c3_w i = 0; i < len_w; ++i) {
+      c3_y c_y = u3r_byte(i, a);
+
+      switch ( c_y ) {
+        case 0 ... 9:
+        case 11 ... 31:
+        {
+          buf_u = _append_text(buf_u, _JSON_UNICODES[c_y], 6);
+          break;
+        }
+
+        case 10:
+        {
+          buf_u = _append_text(buf_u, _JSON_NEWLINE, 2);
+          break;
+        }
+
+        case 34:
+        {
+          buf_u = _append_text(buf_u, _JSON_DOQ, 2);
+          break;
+        }
+
+        case 92:
+        {
+          buf_u = _append_text(buf_u, _JSON_BAS, 2);
+          break;
+        }
+
+        case 127:
+        {
+          buf_u = _append_text(buf_u, _JSON_DEL, 6);
+          break;
+        }
+
+        default:
+        {
+          buf_u = _append_char(buf_u, c_y);
+        }
+      }
+    }
+    buf_u = _append_char(buf_u, '"');
+
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize_array(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+
+    buf_u = _append_char(buf_u, '[');
+
+    if ( u3_nul != a ) {
+      u3_noun i, t = a;
+
+      while ( u3_nul != t ) {
+        u3x_cell(t, &i, &t);
+        buf_u = _serialize(buf_u, i);
+        buf_u = _append_char(buf_u, ',');
+      }
+
+      if ( buf_u == NULL )
+        return NULL;
+      
+      --buf_u->len_w;
+    }
+
+    buf_u = _append_char(buf_u, ']');
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize_object_helper(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+    else if ( u3_nul == a ) {
+      return buf_u;
+    }
+
+    u3_noun n_a, l_a, r_a;
+    u3_noun pn_a, qn_a;
+    u3x_trel(a, &n_a, &l_a, &r_a);
+    u3x_cell(n_a, &pn_a, &qn_a);
+
+    // order is important to match unjetted tree traversal
+    //
+    buf_u = _serialize_object_helper(buf_u, r_a);
+
+    buf_u = _serialize_string(buf_u, pn_a);
+    buf_u = _append_char(buf_u, ':');
+    buf_u = _serialize(buf_u, qn_a);
+    buf_u = _append_char(buf_u, ',');
+
+    buf_u = _serialize_object_helper(buf_u, l_a);
+
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize_object(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+
+    buf_u = _append_char(buf_u, '{');
+
+    if ( u3_nul != a ) {
+      buf_u = _serialize_object_helper(buf_u, a);
+      if ( buf_u == NULL ) {
+        return NULL;
+      }
+      --buf_u->len_w;
+    }
+
+    buf_u = _append_char(buf_u, '}');
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+
+    if ( u3_nul == a ) {
+      return _serialize_null(buf_u);
+    }
+    
+    u3_noun s, p;
+    u3x_cell(a, &s, &p);
+    if ( _(u3du(s)) || _(u3a_is_dog(s)) ) {
+      return _encode_fail(buf_u);
+    }
```

Comment:
This condition should be removed so that unrecognized values fall through to the default case below.

Reply:
Done